Softlayer basic pack

Itxaka opened the pull request:
Softlayer pack with libcloud as backend that allows for create/destroy instance and create/destroy keypair. Tested and working.

I will update this pack in the following weeks with more actions.

Kami commented:
Thanks, I will have a look shortly :)

Kami commented on packs/softlayer/actions/lib/softlayer.py:
```diff
@@ -0,0 +1,36 @@
+            self.logger.error("Node with name {} not found".format(kwargs['name']))
+            exit(1)
+
+    def _run_action(self, **kwargs):
```
Please explicitly declare all the arguments here - it makes it easier to see what's going on.

Kami commented on packs/softlayer/actions/create_instance.yaml:
```diff
@@ -0,0 +1,41 @@
+    immutable: true
+    type: string
+  name:
+    required: true
```
Please add a description for all the non-immutable parameters.

Kami commented:
Added some comments - there are some minor improvements and cleanups which need to be done. Besides that, looks good to me.

Kami commented on packs/softlayer/actions/delete_key_pair.yaml:
```diff
@@ -0,0 +1,13 @@
+    default: delete_key_pair
+    immutable: true
+    type: string
+  key_pair:
```
Looks like this parameter needs to gold a key pair object name, right?

Please clarify that in the parameter description.

Kami commented on packs/softlayer/actions/create_key_pair.yaml:
```diff
@@ -0,0 +1,13 @@
+    default: create_key_pair
+    immutable: true
+    type: string
+  name:
```
Hm, looking at the code, the value of this argument is supposed to be a node hostname?

Or is it supposed to be a node name? or node.name == hostname?

Kami commented on packs/softlayer/actions/lib/softlayer.py:
```diff
@@ -0,0 +1,36 @@
+            return driver.destroy_node(nodes[0])
+        else:
+            self.logger.error("Node with name {} not found".format(kwargs['name']))
+            exit(1)
```
exit takes no arguments, please use `sys.exit`.

Itxaka commented:
@Kami Regarding the parameter names, Im just using the same ones that libcloud expects so I can pass kwargs across and have the functions work "magically" instead of having to check for the method name or have one python file per action.
This is something that was not really clear from st2 documentation. Is there any kind of convention on how to do things? Is one abstract script preferred or one per action? Is explicitly preferred or code simplicity?

Kami commented:
@Itxaka In general, we prefer to be explicit and have multiple action files (one per action), but the way you have it is also fine in this case.

As far as using kwargs goes - the problem with that is that it's really hard to debug and for other users which are not familiar with Libcloud or don't want to go check Libcloud's code to see what's going on.

Explicitly defining arguments does mean a bit more (maintenance) work, but it's easier to follow and see what's going on.

Itxaka commented:
@Kami Perfect! I thought checking other packs that brevity was preferred but I rather have separate files so I have more control onver the actions. Will redo the patch soon!

Itxaka commented:
@Kami I updated the PR.
Basically I changed from the abstract "everything is done magically" to a more decent approach of one entrypoint per action, with proper and easy args and straighforward actions.

Looks much better this time IMHO, so please let me know any issues you can see with it. :+1: 

BTW, I still dont see a lot on the docs on how to react to exceptions or failures in the actions. Do we just let them raise whatever error? Do we try to catch them and return a message with the traceback?

Kami commented:
@Itxaka Thanks, I will have a look shortly.

As far as exceptions go - if they are fatal, in the Python runner, you can just let them propagate / bubble up. Any uncaught exception will cause action to fail (status = failed) and the exception will be reported to the user.

Itxaka commented:
@Kami Perfect regarding the errors. I was waiting for that to make the last commit as I wanted to return a  better error than an IndexErrorr :smile:

Kami commented on packs/softlayer/README.md:
```diff
@@ -0,0 +1,17 @@
+
+* ``username`` - Softlayer username.
+* ``api_key`` - Api key.
+* ``region`` - Not used yet, should be the default region you want to work with for instances.
```
Later on when you add support for other APIs it would make sense to split configuration into `compute` and `storage` section (similar as Libcloud pack does it).

Kami commented on packs/softlayer/actions/create_instance.yaml:
```diff
@@ -0,0 +1,47 @@
+  cpus:
+    required: false
+    type: integer
+    description: Cpus for this instance (defaults to 1).
```
Good practice it to also set the same default value in the action metadata file for all the parameters with default values.

This way this default value is also displayed when running `st2 action -h`.

Itxaka replied:
Good point, updating!

Kami commented:
Merged into master. Thanks.
